### PR TITLE
Generate debug and release Gecko bindings.

### PR DESCRIPTION
**Pull request by heycam**
This makes `generate_style_structs.py` generate bindings for both `DEBUG` and non-`DEBUG` Gecko configurations.  The corresponding generated bindings are used in geckolib based on `cfg(debug-assertions)`.

I regenerated bindings on top of Gecko master plus https://bugzilla.mozilla.org/show_bug.cgi?id=1277154 (landing soon), which tripped up the bindgen due to bitfield layout issues.

r? @emilio 

---

- [ ] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [ ] These changes fix #__ (github issue number if applicable).

- [ ] There are tests for these changes OR
- [X] These changes do not require tests because they are geckolib only

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/11549)

### CODE REVIEW

**Comment by metajack** [comment]
Is there are reason not to just build the bindings for the current build configuration? cargo already switches between these. I suppose the only real downside is making the build time longer, so maybe it's not a big deal.

**Comment by heycam** [comment]
I guess we could.  I've been doing a lot of switching between debug/release configurations, though, and the bindings generation doesn't happen as part of the normal build process, so it's slightly more convenient to have both there.  If the bindings needed to be regenerated due to Gecko changes, it would mean we need to do a `./mach build-geckolib`, run the bindgen tool, `./mach build-geckolib -r`, run the bindgen tool again.

**Comment by metajack** [comment]
Oh, if the bindings aren't generated as part of a build script, then I think this is totally fine. Especially since you're checking in the result now that I look at it, and you'll definitely want to have both if you're doing that.

File: ports/geckolib/gecko_bindings/tools/regen_style_structs.sh [hunk 1]
```diff
@@ -38,99 +38,103 @@ fi
...
-else
-  echo -e "\e[34minfo:\e[0m bindgen exited successfully, running tests"
-  TESTS_FILE=$(mktemp)
-  rustc ../structs.rs --test -o $TESTS_FILE
+echo -e "\e[34minfo:\e[0m bindgen exited successfully, running tests"
+TESTS_FILE=$(mktemp)
+for target in debug release; do
+  rustc ../structs_${target}.rs --test -o $TESTS_FILE
   $TESTS_FILE
   rm $TESTS_FILE
```
Comment (emilio) [inline ports/geckolib/gecko_bindings/tools/regen_style_structs.sh]: nit: This probably should remove the file outside of the loop.

**Comment by emilio** [comment]
LGTM with that nit addressed.

Also, I've noticed that this contained @bholley's overflow-wrap patch, which is also in #11456. I'm fine waiting for this to merge and rebasing my PR if you think this is going to be tedious to rebase.

**Comment by emilio** [comment]
Well, #11456 is in the queue, I can rebase this if needed, but I thought it's nice to preserve the ownership of that commit :-)

**Comment by emilio** [comment]
Regarding the bitfields (just read the bug), I think we can do well enough with bindgen's bitfield methods.

These were disabled at the first time because bindgen had bugs with enums in bitfields, so it generated invalid code.

This was eventually fixed, so I enabled them to access `nsTArray<T>::mCapacity`, for example in the linear-gradient PR.

If you prefer to land that gecko patch though, it's fine.

**Comment by heycam** [comment]
I think it's still a good idea to avoid bitfields since as far as I can tell, their layout and alignment is completely implementation dependent.

**Comment by emilio** [comment]
Fair enough, though though at least clang and gcc agree with its
representation and packing.

**Comment by emilio** [comment]
@bors-servo: r+

So gecko keeps a bunch of `_mOwningThread` in release? I found that surprising. Anyways, I manually verified that some debug-only things like `mImageTracked` don't appear, so looks fine.

**Comment by heycam** [comment]
Hmm, that is surprising to me too.  Can you make bors hold off on landing it while I investigate?

**Comment by emilio** [comment]
Sure,

@bors-servo: r-

**Comment by emilio** [comment]
Feel free to re-land if I'm sleep and the changes aren't too breaking (like adding a flag or something).

@bors-servo: delegate+

**Comment by emilio** [comment]
Possibly it depends on the contents of `mozilla-config.h`. Reading those I get a lot of `#define MOZ_DEBUG_XXX 1`, so maybe this approach won't scale well and we'll have to switch to something like `./regen_style_structs.sh <target>` ensuring gecko build is properly configured.

**Comment by heycam** [comment]
It looks like in nsISupportsImp.h for NIGHTLY_BUILD we still generate the _mOwningThread fields.  However I'm getting crashes while trying to build/run release locally.  This was working for me before I rebased this morning so I might've messed something up.

**Comment by emilio** [comment]
Yeah, I was just reading that `#ifdef` :P

I have to sleep now, but let me know if you want me to take a try to build it with your patch tomorrow or this weekend and report results :)

**Comment by heycam** [comment]
The issue with stylo crashing was another commit f4b8f9766cbcdf26301939fd0993b2b6f90675c5.  So this is good to land.

@bors-servo r+

**Comment by heycam** [comment]
The bindings won't compile without using a nightly rustc so we may as well land 2adc9ab at the same time.

**Pull request merged**